File: src/main/java/me/nasukhov/TukitaLearner/bot/bridge/tg/TelegramOutput.java
package me.nasukhov.TukitaLearner.bot.bridge.tg;

import me.nasukhov.TukitaLearner.bot.io.Output;
import org.telegram.telegrambots.meta.api.methods.groupadministration.GetChatMember;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
import org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class TelegramOutput implements Output {
    private static final Pattern PLACEHOLDER_USERNAME = Pattern.compile("<user>(-?\\d+)</user>");
    private static final Pattern PLACEHOLDER_SPOILER = Pattern.compile("<spoiler>(.+?)</spoiler>");
    private static final Map<Long, String> userNames = new ConcurrentHashMap<>();

    private final Long chatId;
    private final Telegram api;

    public TelegramOutput(Long chatId, Telegram tg) {
        this.chatId = chatId;
        api = tg;
    }

    @Override
    public void write(String text) {
        SendMessage message = new SendMessage();
        message.setChatId(chatId.toString());
        message.setText(renderText(text));
        message.disableNotification();
        message.setParseMode("HTML");
        try {
            api.execute(message);
        } catch (TelegramApiException e) {
            if (handleError(e)) {
                return;
            }
            // TODO
            e.printStackTrace();
        }
    }

    @Override
    public void promptChoice(String question, Map<String, String> replyOptions) {
        SendMessage message = new SendMessage();
        message.setChatId(chatId.toString());
        message.setText(question);
        message.setReplyMarkup(createOptions(replyOptions));
        message.disableNotification();
        try {
            api.execute(message);
        } catch (TelegramApiException e) {
            if (handleError(e)) {
                return;
            }
            // TODO
            e.printStackTrace();
        }
    }

    private InlineKeyboardMarkup createOptions(Map<String, String> options) {
        InlineKeyboardMarkup inlineKeyboard = new InlineKeyboardMarkup();
        List<List<InlineKeyboardButton>> rowsInline = new ArrayList<>();
        for (Map.Entry<String, String> entry : options.entrySet()) {
            InlineKeyboardButton button1 = new InlineKeyboardButton();
            button1.setText(entry.getKey());
            // allows only 64bytes
            button1.setCallbackData(entry.getValue());

            // One button per line
            List<InlineKeyboardButton> rowInline = new ArrayList<>();
            rowInline.add(button1);
            rowsInline.add(rowInline);
        }

        inlineKeyboard.setKeyboard(rowsInline);

        return inlineKeyboard;
    }

    // TODO move to bb-codes rendering classes
    private String renderText(String text) {
        // Id-to-Name renderer
        Matcher matcher = PLACEHOLDER_USERNAME.matcher(text);
        StringBuilder result = new StringBuilder();
        while (matcher.find()) {
            Long userId = Long.parseLong(matcher.group(1));

            Optional<String> name = getUsername(userId);
            if (name.isEmpty()) {
                // This effectively means user had left the channel
                matcher.appendReplacement(result, "Unknown");
            } else {
                matcher.appendReplacement(result, name.get());
            }
        }
        matcher.appendTail(result);

        // Spoiler renderer
        Matcher spoilerMatcher = PLACEHOLDER_SPOILER.matcher(result.toString());
        StringBuilder result2 = new StringBuilder();
        while (spoilerMatcher.find()) {
            spoilerMatcher.appendReplacement(result2, "<span class=\"tg-spoiler\">" + spoilerMatcher.group(1) + "</span>");
        }
        spoilerMatcher.appendTail(result2);

        return result2.toString();
    }

    private Optional<String> getUsername(Long userId) {
        if (userNames.containsKey(userId)) {
            return Optional.of(userNames.get(userId));
        }

        GetChatMember command = new GetChatMember();
        command.setChatId(chatId);
        command.setUserId(userId);
        try {
            String name = api.execute(command).getUser().getFirstName();
            userNames.put(userId, name);

            return Optional.of(name);
        } catch (TelegramApiException e) {
            return Optional.empty();
        }
    }

    private boolean handleError(TelegramApiException error) {
        if (!(error instanceof TelegramApiRequestException)) {
            return false;
        }

        // TODO SRP violation event dispatch event so handler deactivates it
        int errorCode = ((TelegramApiRequestException) error).getErrorCode();
        // Both cases mean that either bot no longer has access to chat
        if (errorCode != 400 && errorCode != 403) {
            return false;
        }

        // TODO IOResolver.telegramChannel(chatId, true).deactivate();

        return true;
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/bot/bridge/tg/TelegramOutput.kt
package me.nasukhov.tukitalearner.bot.bridge.tg

import me.nasukhov.tukitalearner.bot.io.Output
import org.slf4j.LoggerFactory
import org.telegram.telegrambots.meta.api.methods.groupadministration.GetChatMember
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton
import org.telegram.telegrambots.meta.exceptions.TelegramApiException
import org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Matcher
import java.util.regex.Pattern

class TelegramOutput(
    private val chatId: Long,
    private val api: Telegram,
) : Output {
    // TODO KotlinLogging.logger{} is direly incompatible with my intellij idea
    // research what's going on there
    private val logger = LoggerFactory.getLogger(TelegramOutput::class.java)

    override fun write(text: String) {
        val message = SendMessage()
        message.chatId = chatId.toString()
        message.text = renderText(text)
        message.disableNotification()
        message.parseMode = "HTML"
        try {
            api.execute(message)
        } catch (e: TelegramApiException) {
            if (handleError(e)) {
                return
            }
            logger.error(e.message, e)
        }
    }

    override fun promptChoice(
        question: String,
        replyOptions: Map<String, String>,
    ) {
        val message = SendMessage()
        message.chatId = chatId.toString()
        message.text = question
        message.replyMarkup = createOptions(replyOptions)
        message.disableNotification()
        try {
            api.execute(message)
        } catch (e: TelegramApiException) {
            if (handleError(e)) {
                return
            }
            logger.error(e.message, e)
        }
    }

    private fun createOptions(options: Map<String, String>): InlineKeyboardMarkup {
        val inlineKeyboard = InlineKeyboardMarkup()
        val rowsInline = ArrayList<List<InlineKeyboardButton>>()
        for (entry in options.entries) {
            val button1 = InlineKeyboardButton()
            button1.text = entry.key
            // allows only 64bytes
            button1.callbackData = entry.value

            // One button per line
            val rowInline = ArrayList<InlineKeyboardButton>()
            rowInline.add(button1)
            rowsInline.add(rowInline)
        }

        inlineKeyboard.keyboard = rowsInline

        return inlineKeyboard
    }

    // TODO move to bb-codes rendering classes
    private fun renderText(text: String): String {
        // Id-to-Name renderer
        val matcher: Matcher = PLACEHOLDER_USERNAME.matcher(text)
        val result = StringBuilder()
        while (matcher.find()) {
            val userId = matcher.group(1).toLong()

            val name = getUsername(userId)
            if (name.isEmpty) {
                // This effectively means user had left the channel
                matcher.appendReplacement(result, "Unknown")
            } else {
                matcher.appendReplacement(result, name.get())
            }
        }
        matcher.appendTail(result)

        // Spoiler renderer
        val spoilerMatcher: Matcher = PLACEHOLDER_SPOILER.matcher(result.toString())
        val result2 = StringBuilder()
        while (spoilerMatcher.find()) {
            spoilerMatcher.appendReplacement(
                result2,
                "<span class=\"tg-spoiler\">" + spoilerMatcher.group(1) + "</span>",
            )
        }
        spoilerMatcher.appendTail(result2)

        return result2.toString()
    }

    private fun getUsername(userId: Long): Optional<String> {
        if (userNames.containsKey(userId)) {
            return Optional.of(userNames[userId]!!)
        }

        val softMaxSize = 10000
        // Just a safe-switch to prevent indefinite growth. Halves the map. Though, I'd rather remove stale entries
        if (userNames.size >= softMaxSize) {
            userNames.keys
                .take((userNames.size - softMaxSize) / 2)
                .forEach { userNames.remove(it) }
        }

        val command = GetChatMember()
        command.setChatId(chatId)
        command.userId = userId
        try {
            val name = api.execute(command).user.firstName
            userNames.put(userId, name)

            return Optional.of(name)
        } catch (_: TelegramApiException) {
            return Optional.empty()
        }
    }

    private fun handleError(error: TelegramApiException): Boolean {
        if (error !is TelegramApiRequestException) {
            return false
        }

        // TODO SRP violation event dispatch event so handler deactivates it
        val errorCode = error.errorCode
        // Both cases mean that either bot no longer has access to chat
        if (errorCode != 400 && errorCode != 403) {
            return false
        }

        // TODO IOResolver.telegramChannel(chatId, true).deactivate();
        return true
    }

    companion object {
        private val PLACEHOLDER_USERNAME: Pattern = Pattern.compile("<user>(-?\\d+)</user>")
        private val PLACEHOLDER_SPOILER: Pattern = Pattern.compile("<spoiler>(.+?)</spoiler>")
        private val userNames = ConcurrentHashMap<Long, String>()
    }
}
